### ISSUE TITLE
Support Postgres ltree In Prisma.

### ISSUE THREAD

**Issue by aakashkag** [prisma/prisma#2568]
First of all thank you so much to Prisma team for greate work
Many project oftan require to deal with Tree like structure data. 
It will be greate if Prisma can support Postgres Ltree. 

Currently we can use alternate way to store tree like data into Postgres but it's create N+1 issue during read operation.

Reference: [https://dzone.com/articles/trying-to-represent-a-tree-structure-using-postgre](https://dzone.com/articles/trying-to-represent-a-tree-structure-using-postgre)

**Comment by pantharshit00** [comment]
This will most likely be support via #446 in the future

**Comment by trompx** [comment]
Would also benefit from ltree type support for a current project. Do you have an ETA for an eventual implementation or you advice to use typeorm/mikroorm that already support it?

**Comment by halfmatthalfcat** [comment]
> Would also benefit from ltree type support for a current project. Do you have an ETA for an eventual implementation or you advice to use typeorm/mikroorm that already support it? Or maybe until this is implemented, use only raw query and add the ltree column outside of the prisma schema?

@trompx implementation is currently underway. You can follow along in the current discussion here: https://github.com/prisma/prisma/discussions/12306#discussioncomment-2351839

There’s quite a few places/libraries that needed to be changed to support ltree but hopefully we’ll have support landed in the near future.

**Comment by trompx** [comment]
Thanks for the prompt answer @halfmatthalfcat
If by implementing ltree, you mean create a whole layer (like you did for the full-text search), I understand it is going to take some time.
In the meantime, I would prefer to go with prisma, and plan to use raw queries (so I would not need the whole ltree integration in prisma), I would just need to somehow create the ltree type db column.
Would you have a best practice to handle type fields outside of prisma schema, should I create a migration that adds that ltree type db column? Would you have an example by any chance?
Thank you

**Comment by YandiBanyuKarimaWaly** [comment]
> Would you have a best practice to handle type fields outside of prisma schema, should I create a migration that adds that ltree type db column? Would you have an example by any chance?

@trompx  I believe this is covered in the documentation. [See this page.](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#:~:text=Unsupported%20types&text=The%20Unsupported%20attribute%20allows%20you,(%22polygon%22)%20type.)

tl;dr : Use `unsupported("type_name")` to create a column with unsupported data type.

**Comment by halfmatthalfcat** [comment]
> > Would you have a best practice to handle type fields outside of prisma schema, should I create a migration that adds that ltree type db column? Would you have an example by any chance?
> 
> @trompx I believe this is covered in the documentation. [See this page.](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#:~:text=Unsupported%20types&text=The%20Unsupported%20attribute%20allows%20you,(%22polygon%22)%20type.)
> 
> tl;dr : Use `Unsupported("type_name")` to create a column with unsupported data type.

@YandiBanyuKarimaWaly this actually will not work with ltree because the underlying binary (de)serialization that Prisma uses to communicate with Postgres is not updated to understand ltree yet.

There currently is no way to utilize ltree in the Prisma ecosystem because the underlying Postgres Rust driver, rust-postgres, just received support for ltree as a part of this larger effort.

**Comment by YandiBanyuKarimaWaly** [comment]
> @YandiBanyuKarimaWaly this actually will not work with ltree unfortunately. The underlying binary (de)serialization that Prisma uses to communicate with Postgres is not updated to understand ltree yet.
>
> Even with raw queries, rust-postgres, quaint and prisma-engines (all the libraries that enable prisma to create and execute queries against Postgres) would not know how to handle a column of type ltree.

Oh, that's unfortunate. So I guess for a while I'll just use the recursive approach for creating a tree.

**Comment by trompx** [comment]
Thanks again for the clarifications @halfmatthalfcat.
That's really unfortunate indeed. I'd rather not change my tree implementation for a lack of feature at the orm level.
I had the idea to create two tables, one that has the tree data (like category id and name, handled by prisma so I can have one-to-many, many-to-may with other tables), and the second has the ltree path data (like A.B.C), so I can handle the ltree with another ORM, but it seems this is not trivial either to just exclude one table ([as seen in this issue](https://github.com/prisma/prisma/issues/8864)).

**Comment by trompx** [comment]
Hello @janpio, I was wondering if the [planned PostGIS support](https://prismaio.notion.site/Prisma-Roadmap-50766227b779464ab98899accb98295f?v=171789df05134c719e5be2574e96364e&p=525cf298497b4872a03dc7149f09f22d) would also includes the migration of other types (like ltree) or if it was specific to the geospatial types?

**Comment by janpio** [comment]
I don't think there is much overlap with GIS/PostGIS and ltree that would make us include ltree support itself into that project - but as we are just starting out, I can not say for sure. But I am not expecting it.

**Comment by tommyzrce** [comment]
Hi I would also need this, @nikolasburk adviced me to comment on this issue to prioritize it so here is my +1.

**Comment by reco** [comment]
Hi there, would also need this... +1

**Comment by kokotom** [comment]
would love to have ltree support as well +1

**Comment by mlesin** [comment]
We also hope that it will be implemented soon +1

**Comment by halfmatthalfcat** [comment]
Just an update from myself, I have the PRs (more or less) ready to get this into Prisma, however I need support from the Prisma team in order to get everything merged (mostly reviewing PRs and such). Specifically:

* https://github.com/prisma/quaint/pull/352
* https://github.com/halfmatthalfcat/prisma-engines/pull/1/commits/ce55c27800f8e78bbd157cd618ad339effdb4e9c

I have dogfooded this locally and things seem to work. There may be one or two rough edges (in regards to the emitted Prisma Client types for ltree) that need to be rectified but that's about it.

I can't commit any more time to this though unless there's a concerted effort by the Prisma team to get ltree merged.

**Comment by janpio** [comment]
We are currently working on PostgreSQL extensions support which would remove the first hurdle around Introspection and Migrations that would then also allow us to properly test everthing else. (Gist indexes are now supported since 4.0.0, so that is already out of the way. Probably there will be follow up changes necessary though when we get to ltree I guess)

**Comment by shahinghasemi** [comment]
+1 For this. I want to use `ltree` data structure in postgresql database. Is there any workaround?

**Comment by kevin1193** [comment]
> postgresqlExtensions

does this works fine?

**Comment by eliasleyton** [comment]
> > postgresqlExtensions
> 
> does this works fine?

Works for Migrations, but doesn't work directly via Prisma

**Comment by kevin1193** [comment]
> postgresqlExtensions

Got it! Thanks for the feedback. Can you put some example also a prisma implementation based on this model?

**Comment by eliasleyton** [comment]
> > postgresqlExtensions
> 
> Got it! Thanks for the feedback. Can you put some example also a prisma implementation based on this model?

You can use raw prisma querys, your don't use prisma API by default. 

This is only for get a prisma schema sincronizad with your db.

**Comment by kaikun213** [comment]
This would be a great feature addition!

Is there any update if this will become part of the prisma client?

**Comment by kyong0612** [comment]
Could you please provide the current status of this feature request? Many users are interested in the implementation of PostgreSQL ltree support.

**Comment by jdalegonzalez** [comment]
@ivanvpan You're absolutely right.  Which, frankly, begs the question of whether you should or should not use an ORM when you're really interacting with an SQL resource or just skip prisma and use "sql helpers" instead.
